src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -384,23 +513,114 @@ void ClusterDiscovery::initialUpdate()
             throw Exception(ErrorCodes::KEEPER_EXCEPTION, "Failpoint cluster_discovery_faults is triggered");
     });
 
+    for (auto & path : (*multicluster_discovery_paths))
+    {
+        auto zk = context->getDefaultOrAuxiliaryZooKeeper(path->zk_name);
+
+        zk->createAncestors(path->zk_path);
+        zk->createIfNotExists(path->zk_path, "");
+
+        auto watch_callback = [&path](auto) { path->need_update = true; };
+        zk->getChildrenWatch(path->zk_path, nullptr, watch_callback);
+    }
+
+    findDynamicClusters(clusters_info);
+
     for (auto & [_, info] : clusters_info)
     {
         auto zk = context->getDefaultOrAuxiliaryZooKeeper(info.zk_name);
         registerInZk(zk, info);
-        if (!updateCluster(info))
+        if (!upsertCluster(info))
         {
             LOG_WARNING(log, "Error on initial cluster '{}' update, will retry in background", info.name);
-            clusters_to_update->set(info.name);
+            if (info.zk_root_index)
+                dynamic_clusters_to_update->set(info.name);
+            else
+                clusters_to_update->set(info.name);
         }
+        else if (info.zk_root_index)
+            dynamic_clusters_to_update->set(info.name, false);
     }
+
     LOG_DEBUG(log, "Initialized");
     is_initialized = true;
 }
 
+void ClusterDiscovery::findDynamicClusters(
+    std::unordered_map<String, ClusterDiscovery::ClusterInfo> & info,
+    std::unordered_set<size_t> * unchanged_roots)
+{
+    using namespace std::chrono_literals;
+
+    constexpr auto force_update_interval = 2min;
+
+    size_t zk_root_index = 0;
+    
+    for (auto & path : (*multicluster_discovery_paths))
+    {
+        ++zk_root_index;
```

Comment:
Please, explain why we start index from 1.

Reply:
0 is reserved for "static" clusters.
Dynamic clustres are from Keeper, records in Keeper are created by other nodes during selfregister.
Static clusters are from local xml config.

Reply:
May be reflect it as a comment?

Reply:
There is a comment in ClusterDiscovery.h near zk_root_index.